### ISSUE TITLE
DependencyResolutionException in Windows 10 Store app

### ISSUE THREAD

**Issue by levarius** [autofac/Autofac#686]
After doing a build with the .NET Native tool chain enabled, I am getting a strange DependencyResolutionException when trying to resolve a view model. The viewmodel, and all of the parameters in the constructor have been registered into the container. The resolving happens as expected in a Debug build without the .NET Native tool chain optimizations enabled.

Here is what the constructor looks like:
`public MyViewModel(INavigationService navigationService, ResourceLoader loader, HttpMessageHandler handler, AccountManager accountManager, ISettingsManager settingsManager, IConverter converter, Func<DateTime, Enum, SubViewModel> itemFactory, Enum enum = Enum.Default)`

However, if I decrease the number of parameters pulled from the IoC to 3 (remove the `AccountManager`, `ISettingsManager`, and `IConverter`), the view model can be resolved correctly. Additionally, if I remove the default `Enum` parameter from the end of the constructor, I can leave all of the other parameters alone and it works as expected.

**Comment by tillig** [comment]
Unfortunately to diagnose something like this you'll need to provide the full exception you're seeing as well as a small repro so we can see it in action.

**Comment by levarius** [comment]
Of course. I've included the full stack trace below. I've also added the inner exception, and will post a sample project with the issue shortly.

`{Autofac.Core.DependencyResolutionException: The runtime has refused to evaluate the expression at this time.`
` at Autofac.Core.Activators.Reflection.ConstructorParameterBinding.Instantiate()`
` at Autofac.Core.Activators.Reflection.ReflectionActivator.ActivateInstance(IComponentContext context, IEnumerable'1 parameters)`
` at Autofac.Core.Resolving.InstanceLookup.Activate(IEnumerable'1 parameters)\n   at Autofac.Core.Resolving.InstanceLookup.Execute()`
` at Autofac.Core.Resolving.ResolveOperation.GetOrCreateInstance(ISharingLifetimeScope currentOperationScope, IComponentRegistration registration, IEnumerable'1 parameters)`
` at Autofac.Core.Resolving.ResolveOperation.ResolveComponent(IComponentRegistration registration, IEnumerable'1 parameters)`
` at Autofac.Core.Resolving.ResolveOperation.Execute(IComponentRegistration registration, IEnumerable'1 parameters)`
` at Autofac.Core.Lifetime.LifetimeScope.ResolveComponent(IComponentRegistration registration, IEnumerable'1 parameters)`
` at Autofac.ResolutionExtensions.TryResolveService(IComponentContext context, Service service, IEnumerable'1 parameters, Object& instance)`
` at Autofac.ResolutionExtensions.ResolveService(IComponentContext context, Service service, IEnumerable'1 parameters)`
` at Autofac.ResolutionExtensions.Resolve(IComponentContext context, Type serviceType, IEnumerable'1 parameters)`
` at Autofac.ResolutionExtensions.Resolve(IComponentContext context, Type serviceType)`
` at MyApp.Services.NavigationService.Navigate(Type viewModelType, NavigationArgs args)}`

**Comment by levarius** [comment]
The inner exception:

`{System.ArgumentException: Object of type 'MyApp.Services.INavigationService' cannot be converted to type 'MyApp.ViewModels.MyViewModel'.
   at System.InvokeUtils.CheckArgument(Object srcObject, EETypePtr dstEEType, CheckArgumentSemantics semantics) in f:\dd\ndp\fxcore\src\System.Private.CoreLib\System\InvokeUtils.cs:line 78`
`   at System.InvokeUtils.DynamicInvokeParamHelperCore(RuntimeTypeHandle type, DynamicInvokeParamLookupType& paramLookupType, Int32& index, DynamicInvokeParamType paramType) in f:\dd\ndp\fxcore\src\System.Private.CoreLib\System\InvokeUtils.cs:line 855`
`   at System.InvokeUtils.DynamicInvokeParamHelperIn(RuntimeTypeHandle rth)`
`   at _$ILCT$.$ILT$ReflectionDynamicInvoke$.InvokeRetVIIIIII[T0,T1,T2,T3,T4,T5](Object thisPtr, IntPtr methodToCall, ArgSetupState& argSetupState, Boolean targetIsThisCall)`
`   at System.InvokeUtils.CallIHelperStaticCall(Object thisPtr, IntPtr methodToCall, IntPtr dynamicInvokeHelperMethod, ArgSetupState& argSetupState, Boolean isTargetThisCall)`
`   at System.InvokeUtils.CallDynamicInvokeMethod(Object thisPtr, IntPtr methodToCall, Object thisPtrDynamicInvokeMethod, IntPtr dynamicInvokeHelperMethod, IntPtr dynamicInvokeHelperGenericDictionary, String defaultValueString, Object[] parameters, Boolean invokeMethodHelperIsThisCall, Boolean methodToCallIsThisCall) in f:\dd\ndp\fxcore\src\System.Private.CoreLib\System\InvokeUtils.cs:line 497`
`   at Internal.Runtime.Augments.RuntimeAugments.CallDynamicInvokeMethod(Object thisPtr, IntPtr methodToCall, Object thisPtrDynamicInvokeMethod, IntPtr dynamicInvokeHelperMethod, IntPtr dynamicInvokeHelperGenericDictionary, String defaultValueString, Object[] parameters, Boolean invokeMethodHelperIsThisCall, Boolean methodToCallIsThisCall) in f:\dd\ndp\fxcore\src\System.Private.CoreLib\Internal\Runtime\Augments\RuntimeAugments.cs:line 281`
`   at Internal.Reflection.Execution.MethodInvokers.InstanceMethodInvoker.Invoke(Object thisObject, Object[] arguments) in f:\dd\ndp\fxcore\src\System.Private.Reflection.Execution\Internal\Reflection\Execution\MethodInvokers\InstanceMethodInvoker.cs:line 34`
`   at System.Reflection.Runtime.MethodInfos.RuntimePlainConstructorInfo.Invoke(Object[] parameters) in f:\dd\ndp\fxcore\src\System.Private.Reflection.Core\System\Reflection\Runtime\MethodInfos\RuntimePlainConstructorInfo.cs:line 99`
`   at System.Linq.Expressions.Interpreter.NewInstruction.Run(InterpretedFrame frame)`
`   at System.Linq.Expressions.Interpreter.Interpreter.Run(InterpretedFrame frame)`
`   at System.Linq.Expressions.Interpreter.LightLambda.Run(Object[] arguments)`
`   at System.Func'2.Invoke(T arg)`
`   at System.Func'2.ReverseInvokeImpl(T arg)`
`   at System.Func'2.Invoke(T arg)`
`   at Autofac.Core.Activators.Reflection.ConstructorParameterBinding.Instantiate()}`

**Comment by tillig** [comment]
Not that it solves the problem, but from a "keeping notes on the research" perspective, searching on the error message "The runtime has refused to evaluate the expression at this time." brings up a lot of notes about the debugger:

- http://stackoverflow.com/questions/19843683/vs2013-debugger-entity-framework-runtime-has-refused-to-evaluate-the-express
- http://weblog.west-wind.com/posts/2013/Nov/21/Visual-Studio-2013-Could-not-evaluate-Expression-Debugger-Abnormality

I'm also looking at the stack trace and... this is some deep, low-level stuff where a function is being invoked and it looks like something's going wrong with the compiler-generated expression tree. Depending on how the repro goes, it may or may not be something we can fix; it may be something broken lower down.

**Comment by levarius** [comment]
Thanks for those posts. I should have mentioned we are using Visual Studio 2015, and are targeting Windows 10 Universal (Build 10240).

We also had some luck with another external library by adding some Runtime Directives to help with the tree-shaking the .NET Native tool chain is doing. Not sure if those would help here, but I can send you what we discovered.

**Comment by tillig** [comment]
Any tips or pointers are always welcome. Thanks!

**Comment by levarius** [comment]
I can't seem to attach the zip of the project. Is there a better way for me to get the sample to you?

Once opened, the app will currently crash if you run with .NET Native tool chain compilation enabled. There are 2 ways to fix this:

* Remove the default Enum parameter at the end of the parameter list.
* Remove 3+ parameters in front of the Func in the parameter list.

The Runtime DIrectives we added to help with keeping necessary components around after tree-shaking are documented here: http://go.microsoft.com/fwlink/?LinkID=391919

Here is a sample of what we added to our Default.rd.xml file:
```xml
<Directives xmlns="http://schemas.microsoft.com/netfx/2013/01/metadata">
	<Application>
		<Assembly Name="*Application*" Dynamic="Required All" />

		<!-- Add your application specific runtime directives here. -->
		<Type Name="Telerik.Core.BindingExpressionHelper">
			<MethodInstantiation Name="ToUntypedFunc" Arguments="Garmin.Connect.Net.Models.MetricsData,System.Nullable&lt;System.Double&gt;" Dynamic="Required" />
			<MethodInstantiation Name="ToUntypedFunc" Arguments="Garmin.Connect.Net.Models.MetricsData,System.Nullable&lt;System.DateTime&gt;" Dynamic="Required" />
		</Type>
	</Application>
</Directives>
```

**Comment by tillig** [comment]
Can you post it as a GitHub repo I can just check out?

I have to admit I'm not at all familiar with .NET Native so reading at the link you sent was helpful. Reading about the runtime directives makes me think it would be worth adding directives that... well, basically just opened _literally everything in the application_ up for full reflection and constructor instantiation. Autofac was written with the assumption it's in a runtime that allows for that (IL/CLR); the runtime directives appear to be required to open that up.

Just for experimentation, I might try adding all of the assemblies in the app to the list, like

    <Assembly Name="TheAssembly"
              Activate="Required All"
              Browse="Required All"
              MarshalObject="Required All"
              MarshalDelegate="Required All"
              MarshalStructure="Required All" />

Basically, just crank open the floodgates to see if it even makes a difference. Looking at the list of policies, I don't think Autofac would need Dynamic or any of the serialization policies since it doesn't do dynamic programming or any serialization. It does, however, do reflection inspection on things, activates objects on the fly, and passes references around. (I'm not entirely sure the marshaling policies will help, but, you know, just thinking _open it all up and see what happens_.)

It appears you might be able to set this stuff on the entire application by putting the policies on the `<Application>` element directly.

**Comment by levarius** [comment]
Uploaded sample project: https://github.com/levarius/autofac-demo

I'll give those directives a try and let you know if anything changes.

**Comment by tillig** [comment]
Any luck with the runtime directives?

**Comment by levarius** [comment]
We haven't had an opportunity to go back and test this fix yet. We found a workaround for now, just removing the optional parameter from the end of the constructor. I'll let you know if/when we are able to check these fixes out.

**Comment by tillig** [comment]
I'm going to close this for now. If it turns out there's actually something Autofac needs to do to make things work, we can reopen.
